### PR TITLE
Add error codes to troubleshooting and a page for GenQuery

### PR DESCRIPTION
**Pull request by alanking**
Addresses irods/irods#6631

I already spent the time on this so I'm putting this up here, but we don't have to use it.

The only thing the issue calls for is the error codes and a page for `iquest attrs`, so, take what seems good and we'll throw out the rest (which could be all of it)

### CODE REVIEW

File: docs/system_overview/genquery.md [hunk 1]
```diff
@@ -0,0 +1,443 @@
+#
+
+## iRODS General Query (GenQuery)
+
+GenQuery is the SQL-like syntax used to query the iRODS catalog. It is used throughout the server to accomplish everything required of the system from resource hierarchy resolution to enforcing permissions, as well as directly by users via `iquest` and other clients.
+
+This page describes syntax for GenQuery as it has existed in iRODS for almost as long as it has existed. A more modern solution is now available in the iRODS server called GenQuery2 which will eventually replace the legacy GenQuery implementation.
```
Comment (korydraughn) [inline docs/system_overview/genquery.md:7]: The part about how long GenQuery has existed reads a little weird to me. Consider rephrasing that part.

**Review by korydraughn — COMMENTED** [review]
I like it. Once the review comments are handled, this will be ready.

File: docs/system_overview/genquery.md [hunk 2]
```diff
@@ -0,0 +1,443 @@
...
+SELECT USER_NAME, SUM(DATA_SIZE), COUNT(DATA_NAME), RESC_NAME
+```
+Find the aggregate size of the data objects in `/tempZone/home/rods` and all of its subcollections for each resource:
+```
+SELECT SUM(DATA_SIZE), RESC_NAME WHERE COLL_NAME LIKE '/tempZone/home/rods%'
+```
+Get a list of data object IDs in `/tempZone/home/rods` and all of its subcollections, ordered from highest ID value to lowest ID value:
+```
+SELECT ORDER_DESC(DATA_ID) WHERE COLL_NAME LIKE '/tempZone/home/rods%'
+```
```
[the remark this answers is not in the document]
Comment (alanking) [reply docs/system_overview/genquery.md:431]: Oh, I see. I took this right out of the `iquest` help text so I haven't even tried any of these lol

I'll try adding other columns. Do you happen to know the effect of adding multiple `order`/`order_desc` selections? Does it order it "in order" of how they are placed?

File: docs/system_overview/genquery.md (hunk 1 above)
Comment (alanking) [reply docs/system_overview/genquery.md:7]: I agree. Must have been tired when I wrote that lol

File: docs/system_overview/genquery.md [hunk 3]
```diff
@@ -0,0 +1,443 @@
...
+ATTRIBUTE is a GenQuery attribute. See [GenQuery Attributes](#genquery-attributes) for a list of attributes.
+
+OPERATOR can be one of a few different relational operators, most of which will be familiar to traditional SQL users:
+ - '=': ATTRIBUTE must exactly match VALUE.
+ - '<>'/'!=': ATTRIBUTE must not match VALUE.
+ - '>'/'>=': ATTRIBUTE must be greater than (or greater than or equal to) VALUE.
+ - '<'/'<=': ATTRIBUTE must be less than (or less than or equal to) VALUE.
+ - 'like': ATTRIBUTE must match the wildcard expression in VALUE. See [Wildcard Expressions](#wildcard-expressions) for more information.
+ - 'not like': ATTRIBUTE must not match the wildcard expression in VALUE. See [Wildcard Expressions](#wildcard-expressions) for more information.
+ - 'between': ATTRIBUTE must be between two VALUEs, separated by a space. For example: `DATA_SIZE between '100' '1000'`
```
[the remark this answers is not in the document]
Comment (alanking) [reply docs/system_overview/genquery.md:52]: Will add

File: docs/system_overview/genquery.md [hunk 4]
```diff
@@ -0,0 +1,443 @@
...
+ - RESC_COMMENT
+ - RESC_CONTEXT
+ - RESC_CREATE_TIME
+ - RESC_FREE_SPACE
+ - RESC_FREE_SPACE_TIME
+ - RESC_ID
+ - RESC_INFO
+ - RESC_LOC
+ - RESC_MODIFY_TIME
+ - RESC_NAME
```
[the remark this answers is not in the document]
Comment (alanking) [reply docs/system_overview/genquery.md:280]: Oh, interesting. I took this straight from `iquest attrs` so I didn't really check. I'll compare the list to see what may be missing. For sure we will include `RESC_MODIFY_TIME_MILLIS`.

File: docs/system_overview/genquery.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (alanking) [reply docs/system_overview/genquery.md:54]: Done

File: docs/system_overview/genquery.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (alanking) [reply docs/system_overview/genquery.md]: I added another column to the example and added some clarifying language near the order operators description under "Selection syntax". I'll link the commit to #97 as well if it seems acceptable.

File: docs/system_overview/genquery.md (hunk 1 above)
Comment (alanking) [reply docs/system_overview/genquery.md]: I reworded the phrase. I'll resolve this but feel free to suggest more tweaks

File: docs/system_overview/genquery.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (alanking) [reply docs/system_overview/genquery.md]: The list I generated here must have come from 4.3.1, hence the missing attributes. The `RESC_GROUP_*` attributes are not in fact exposed by GenQuery, so I need to remove those.

[the remark this answers is not in the document]
Comment (alanking) [reply docs/system_overview/genquery.md]: Done

File: docs/system_overview/genquery.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (korydraughn) [reply docs/system_overview/genquery.md]: > Do you happen to know the effect of adding multiple order/order_desc selections?

I don't remember.

> Does it order it "in order" of how they are placed?

My guess is each column _(wrapped in an order directive)_ acts as a scope for the next  column _(wrapped in an order directive)_. That's assuming the ordering logic is maintained in the generated SQL.

Again, just a theory. It's better to test and see what happens.

**Comment by korydraughn** [comment]
Let's include it just for completeness. You can document the syntax and include a very brief statement about it being the equivalent of `where CONDITIONS and (ATTRIBUTE = 'V1' or ATTRIBUTE = 'V2')`.

**Comment by alanking** [comment]
Okay, I added the `||` piece. I'll take a look into the `order`/`order_desc` thing to see if I can draw any conclusions.

File: docs/system_overview/genquery.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (alanking) [reply docs/system_overview/genquery.md]: Okay, after some testing I think we have a better understanding of how things work in GenQuery. I've recorded the behavior even though it's... less than ideal. I don't think we're going to change it at this point, though.

I also expanded the aggregation operator section. Resolving this.

**Comment by alanking** [comment]
I will squash this soon (no #s) if there are no further comments.

File: docs/system_overview/genquery.md [hunk 5]
```diff
@@ -0,0 +1,550 @@
...
+```
+Any results with `DATA_NAME` values that match any of the 3 conditions will be returned. Note that these conditions are checked in order, so it is recommended to use ascending order of specificity for efficiency.
+
+### Wildcard Expressions
+
+VALUE must be surrounded by single-quotes and should be a constant value or wildcard expression. The following wildcard operators are supported:
+ - '%': Matches on any number of any characters (including nothing). This is equivalent to '.\*' in traditional regular expressions.
+ - '\_': Matches on any single character. This is equivalent to '.' in traditional regular expressions.
+
+If '%' or '\_' must be used as part of the value for the query (rather than in a wildcard expression), these can be escaped using a backslash (\\).
```
Comment (trel) [inline docs/system_overview/genquery.md:162]: consider...

'If' -> 'If a literal'

**Comment by korydraughn** [comment]
The suggestions by @trel seem good. Other than that, the PR is ready in my book.

File: docs/system_overview/genquery.md (hunk 5 above)
Comment (alanking) [reply docs/system_overview/genquery.md]: Done

**Comment by korydraughn** [comment]
Looks good. Let's squash it.

File: docs/system_overview/genquery.md [hunk 6]
```diff
@@ -0,0 +1,550 @@
...
+
+OPERATOR can be one of a few different relational operators, most of which will be familiar to traditional SQL users:
+ - '=': ATTRIBUTE must exactly match VALUE.
+ - '<>'/'!=': ATTRIBUTE must not match VALUE.
+ - '>'/'>=': ATTRIBUTE must be greater than (or greater than or equal to) VALUE.
+ - '<'/'<=': ATTRIBUTE must be less than (or less than or equal to) VALUE.
+ - 'like': ATTRIBUTE must match the wildcard expression in VALUE. See [Wildcard Expressions](#wildcard-expressions) for more information.
+ - 'not like': ATTRIBUTE must not match the wildcard expression in VALUE. See [Wildcard Expressions](#wildcard-expressions) for more information.
+ - 'between': ATTRIBUTE must be between two VALUEs, separated by a space. For example: `DATA_SIZE between '100' '1000'`
+ - 'in': ATTRIBUTE must exactly match at least one VALUE in a comma-delimited list of VALUEs. For example: `RESC_NAME in 'resc_a', 'resc_b'`
```
[the remark this answers is not in the document]
Comment (alanking) [reply docs/system_overview/genquery.md]: Yes, I've actually never seen the `in` operator used *without* parentheses before I wrote this page, so I know that with parentheses works.

[the remark this answers is not in the document]
Comment (alanking) [reply docs/system_overview/genquery.md]: From 4.3.1...
```bash
$ iquest "select DATA_ID where DATA_NAME in 'foo', 'goo'"
DATA_ID = 10353
------------------------------------------------------------
DATA_ID = 10354
------------------------------------------------------------
$ iquest "select DATA_ID where DATA_NAME in ('foo', 'goo')"
DATA_ID = 10353
------------------------------------------------------------
DATA_ID = 10354
------------------------------------------------------------
```

**Comment by alanking** [comment]
Squashed.

**Review by korydraughn — APPROVED** [review]
Pound it.

**Comment by alanking** [comment]
#'d. Mergin.

**Pull request merged**